segfault when running with memory limit but without swap limit

**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

Explicitly removing swap limit causes a segmentation fault.

**Steps to reproduce the issue:**

1. `sudo podman run --memory 256m --memory-swap -1 busybox true`

2.

3.

**Describe the results you received:**

`panic: runtime error: invalid memory address or nil pointer dereference`


**Describe the results you expected:**


**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
podman version 3.0.0
```

**Output of `podman info --debug`:**

```
(paste your output here)
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
(paste your output here)
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

No

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Ubuntu 20.04

----

These are supposed to be equivalent:
```console
$ sudo podman run --memory 256m busybox true
Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
$ sudo podman run --memory 256m --memory-swap -1 busybox true
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x15ee61e]

goroutine 1 [running]:
github.com/containers/podman/v2/cmd/podman/common.getMemoryLimits(0xc00034ec00, 0x27d8da0, 0x1b05550, 0x17, 0xc0005bc508)
	cmd/podman/common/specgen.go:153 +0x33e
github.com/containers/podman/v2/cmd/podman/common.FillOutSpecGen(0xc00034ec00, 0x27d8da0, 0xc000139600, 0x4, 0x8, 0x0, 0x18ab660)
	cmd/podman/common/specgen.go:430 +0xda6
github.com/containers/podman/v2/cmd/podman/containers.run(0x2748b80, 0xc000139600, 0x4, 0x8, 0x0, 0x0)
	cmd/podman/containers/run.go:171 +0x46f
github.com/spf13/cobra.(*Command).execute(0x2748b80, 0xc00003c0c0, 0x8, 0x8, 0x2748b80, 0xc00003c0c0)
	vendor/github.com/spf13/cobra/command.go:850 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0x275c140, 0xc000044088, 0x18d6c80, 0x280b340)
	vendor/github.com/spf13/cobra/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	vendor/github.com/spf13/cobra/command.go:895
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	vendor/github.com/spf13/cobra/command.go:888
main.Execute()
	cmd/podman/root.go:92 +0xec
main.main()
	cmd/podman/main.go:36 +0x92
```
System doesn't have "cgroup_enable=memory swapaccount=1" cmdline

--- a/cmd/podman/common/specgen.go
+++ b/cmd/podman/common/specgen.go
@@ -148,17 +148,16 @@ func getMemoryLimits(s *specgen.SpecGenerator, c *ContainerCLIOpts) (*specs.Linu
 	}
 	if m := c.MemorySwap; len(m) > 0 {
 		var ms int64
-		if m == "-1" {
-			ms = int64(-1)
-			s.ResourceLimits.Memory.Swap = &ms
-		} else {
+		// only set memory swap if it was set
+		// -1 indicates unlimited
+		if m != "-1" {
 			ms, err = units.RAMInBytes(m)
+			memory.Swap = &ms
 			if err != nil {
 				return nil, errors.Wrapf(err, "invalid value for memory")
 			}
+			hasLimits = true
 		}
-		memory.Swap = &ms
-		hasLimits = true
 	}
 	if m := c.KernelMemory; len(m) > 0 {
 		mk, err := units.RAMInBytes(m)

--- a/test/e2e/run_memory_test.go
+++ b/test/e2e/run_memory_test.go
@@ -2,6 +2,7 @@ package integration
 
 import (
 	"os"
+	"strconv"
 
 	. "github.com/containers/podman/v2/test/utils"
 	. "github.com/onsi/ginkgo"
@@ -89,5 +90,28 @@ var _ = Describe("Podman run memory", func() {
 		session.WaitWithDefaultTimeout()
 		Expect(session.ExitCode()).To(Equal(0))
 		Expect(session.OutputToString()).To(Equal("41943040"))
+	})
+
+	It("podman run kernel-memory test", func() {
+		if podmanTest.Host.Distribution == "ubuntu" {
+			Skip("Unable to perform test on Ubuntu distributions due to memory management")
+		}
+		var session *PodmanSessionIntegration
+		if CGROUPSV2 {
+			session = podmanTest.Podman([]string{"run", "--memory", "256m", "--memory-swap", "-1", ALPINE, "cat", "/sys/fs/cgroup/memory.swap.max"})
+		} else {
+			session = podmanTest.Podman([]string{"run", "--cgroupns=private", ALPINE, "cat", "/sys/fs/cgroup/memory/memory.memsw.limit_in_bytes"})
+		}
+		session.WaitWithDefaultTimeout()
+		Expect(session.ExitCode()).To(BeZero())
+		output := session.OutputToString()
+		Expect(err).To(BeNil())
+		if CGROUPSV2 {
+			Expect(output).To(Equal("max"))
+		} else {
+			crazyHighNumber, err := strconv.ParseInt(output, 10, 64)
+			Expect(err).To(BeZero())
+			Expect(crazyHighNumber).To(BeNumerically(">", 936854771712))
+		}
 	})
 })

